### PR TITLE
Rate-limit initial seed peer connections

### LINKED ISSUE #2326: Security: Rate-limit initial seed peer connections
## Motivation

Zebra doesn't rate-limit its connections to initial seed peers. There can be more than 100 of these peers on mainnet.

This is a medium-priority security fix, because initial peer connections only happen at Zebra startup.

## Suggested Design

In `add_initial_peers`, make sure only one connection is initiated per `MIN_PEER_CONNECTION_INTERVAL`. This applies the rate limit before Zebra does any work.

`MIN_PEER_CONNECTION_INTERVAL` is the existing rate-limit for outbound `CandidateSet` connections.

### Edge Cases

At the moment, the futures in `add_initial_peers` run concurrently, so just adding `tokio::time::sleep(MIN_PEER_CONNECTION_INTERVAL)` to each future won't work.

Once a connection has been initiated, we need to run it concurrently with all the other connections that are in progress. Otherwise, Zebra's startup will be very slow.

We'd like to make the first connection as soon as `add_initial_peers` is called. This is a low priority, because the interval is short.

### Fragile Tests

This ticket might increase the number of CI failures, or make some acceptance tests fail.
As a temporary fix, try increasing the acceptance test timeouts, or merging ticket #2900.

CI fails with errors like:

> Oct 15 18:27:56.622  INFO {zebrad="4648f8f" net="Main"}: zebra_network::peer_set::candidate_set: timeout waiting for the peer service to become ready

https://github.com/ZcashFoundation/zebra/runs/3909044730#step:10:1246

## Related Work

These tickets modify the same functions as this ticket:
- #2900

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

See https://github.com/ZcashFoundation/zebra/issues/2326

## Solution

This was a first attempt, I'm not sure if it makes sense of if there is a simpler way to do it.

Closes https://github.com/ZcashFoundation/zebra/issues/2326

## Review

Just a draft of the idea I had. @teor2345 and/or @jvff might want to take a look.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Review by teor2345 — COMMENTED** [review]
This looks like a good solution, but the code is a bit complex, so it needs a test.

Here's one way we could test it:
1. Create a config with `initial_mainnet_peers` set to some fake unique IP addresses
2. Call `add_initial_peers` with a fake `connector` and `peerset_tx`
3. Make sure `add_initial_peers` takes long enough

**Comment by teor2345** [comment]
You also asked if there was a simpler way to do it:
1. Assign each peer a sequential number `n` using `enumerate()`
2. Give each peer a timeout of `n * MIN_PEER_CONNECTION_INTERVAL`

(This generates similar timeouts to the current implementation, but without the channels or spawned timer task.)

I don't mind which solution we use, as long as it passes the tests.

**Comment by teor2345** [comment]
> Here's one way we could test it:
>     1. Create a config with `initial_mainnet_peers` set to some fake unique IP addresses
>     2. Call `add_initial_peers` with a fake `connector` and `peerset_tx`

If you want to work on these tests together, I'm going to need something similar for PR #2944.
(Except I'll need to count the number of outbound connections somehow.)

**Comment by conradoplg** [comment]
> You also asked if there was a simpler way to do it:
> 
> 1. Assign each peer a sequential number `n` using `enumerate()`
> 2. Give each peer a timeout of `n * MIN_PEER_CONNECTION_INTERVAL`

Thanks, good idea. I reverted the previous and implemented this in fb17c4399

> If you want to work on these tests together, I'm going to need something similar for PR #2944. (Except I'll need to count the number of outbound connections somehow.)

Sure! I'll think a bit about how to write it and we can discuss it

File: zebra-network/src/peer_set/initialize.rs [hunk 1]
```diff
@@ -233,17 +237,24 @@ where
...
 
-            outbound_connector
-                .clone()
-                .oneshot(req)
-                .map_err(move |e| (addr, e))
+            let outbound_connector = outbound_connector.clone();
+            async move {
+                // Rate-limit the connection, sleeping for an interval according
+                // to its index in the list.
+                sleep(constants::MIN_PEER_CONNECTION_INTERVAL.saturating_mul(i as u32)).await;
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:280]: > saturating_mul

Nice! Overflow will probably never happen, but saturating is better than panicking or wrapping.

**Comment by teor2345** [comment]
> > If you want to work on these tests together, I'm going to need something similar for PR #2944. (Except I'll need to count the number of outbound connections somehow.)
> 
> Sure! I'll think a bit about how to write it and we can discuss it

Here's what I got done today:
https://github.com/ZcashFoundation/zebra/pull/2944#discussion_r736215310

It doesn't quite test the number of connections.
But I did find a task starvation bug using these simple tests.

**Comment by conradoplg** [comment]
> 
> Here's what I got done today: [#2944 (comment)](https://github.com/ZcashFoundation/zebra/pull/2944#discussion_r736215310)
> 
> It doesn't quite test the number of connections. But I did find a task starvation bug using these simple tests.

Thanks!

Added a test in 6db572c76

I cherry-picked 5d245cf but not 0b5f82e since it did not apply cleanly, and ended up not needing it.

Moving out of draft

File: zebra-network/src/peer_set/initialize/tests/vectors.rs [hunk 2]
```diff
@@ -133,3 +144,78 @@ async fn local_listener_port_with(listen_addr: SocketAddr, network: Network) {
...
+
+    assert_eq!(connections.len(), PEER_COUNT);
+    // Make sure the rate limiting worked by checking if it took long enough
+    assert!(
+        elapsed > constants::MIN_PEER_CONNECTION_INTERVAL.saturating_mul((PEER_COUNT - 1) as u32),
+        "elapsed only {:?}",
+        elapsed
+    );
+
+    let _ = crawl_task_handle.await;
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize/tests/vectors.rs]: Test Coverage: Let's also check for errors or panics in the initial peers task.

(If we ignore the task result, the tests will pass even if there are panics in the task, or errors returned by the task.)

I also fixed the name of the `JoinHandle` to match the task we're spawning.

```suggestion
    let (initial_peers_task_handle, peerset_rx) = spawn_add_initial_peers(PEER_COUNT, outbound_connector);
    let connections = peerset_rx.take(PEER_COUNT).collect::<Vec<_>>().await;

    let elapsed = Instant::now() - before;

    assert_eq!(connections.len(), PEER_COUNT);
    // Make sure the rate limiting worked by checking if it took long enough
    assert!(
        elapsed > constants::MIN_PEER_CONNECTION_INTERVAL.saturating_mul((PEER_COUNT - 1) as u32),
        "elapsed only {:?}",
        elapsed
    );

    let initial_peers_result = initial_peers_task_handle.await;
    assert!(
        matches!(initial_peers_result, Ok(Ok(_)))
        "unexpected error or panic in add_initial_peers task: {:?}",
        initial_peers_result,
    );
```

**Review by teor2345 — COMMENTED** [review]
Let's also check for panics or errors in the initial peers task.

File: zebra-network/src/peer_set/initialize/tests/vectors.rs (hunk 2 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize/tests/vectors.rs]: I pushed a commit with this fix to the PR so we could merge.

**Review by teor2345 — DISMISSED** [review]
I applied a suggestion that checks for extra errors in the test.

**Comment by conradoplg** [comment]
> I applied a suggestion that checks for extra errors in the test.

Thanks!

I've synced and solved conflicts so this needs a re-approval

File: zebra-network/src/peer_set/initialize.rs [hunk 3]
```diff
@@ -478,6 +492,11 @@ where
             // Zebra can't control how many queued connections are waiting,
             // but most OSes also limit the number of queued inbound connections on a listener port.
             tokio::time::sleep(constants::MIN_PEER_CONNECTION_INTERVAL).await;
+        } else {
+            // Security: Let other tasks run after each errored connection is processed.
+            //
+            // Avoids remote peers starving other Zebra tasks using inbound connection errors.
+            tokio::task::yield_now().await;
         }
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs]: Remove duplicate code: I made the yield unconditional in the PR I just merged.

So I'll just merge this fix now.
```suggestion
        }
```

**Review by teor2345 — APPROVED** [review]
Looks good, let's merge!

**Pull request merged**